CIP-XXXX/README.md

```diff
@@ -0,0 +1,270 @@
+Because there are less than 256 built-ins, this is currently the same
+as hashing the prefix byte prepended to the byte containing the index of
+the built-in.
+
+`Var` is handled the exact same way (with a different prefix),
+but it's in this case feasible for the index to be more than 255.
+
+In pseudocode: `hash $ prefix <> serialiseBase256 index`
+
+### Hashing `Apply`, `Force`, `Delay`
+
+These are hashed by hashing the result of prepending the prefix
+byte to the concatenation of the hashes of the children.
+
+In pseudocode: `hash $ foldl' (<>) prefix (hash <$> children)`
+
+### Hashing `LamAbs`
+
+This works the exact same way as above, notably, the _name_ is excluded
+as it's a constant in the de-Bruijn encoding.
+
+In pseudocode: `hash $ prefix <> hash body`
+
+### Hashing `Constant`
+
+The universe of types used on-chain is always `DefaultUni`.
+Each possible data type is handled differently, with each having
+a different prefix. The total number of prefixes does not exceed
+255. If it did, the prefix would have to be increased to two bytes.
+
+In addition:
+Negative integers and non-negative integers have separate prefixes.
+False and True also have separate prefixes.
+
+#### Hashing non-negative integers
+
+The serialisation according to [CIP-58](https://github.com/cardano-foundation/CIPs/blob/a1b9ff0190ad9f3e51ae23e85c7a8f29583278f0/CIP-%3F/README.md#representation-of-builtininteger-as-builtinbytestring-and-conversions),
+prefixed with the two-byte prefix, is hashed.
+
+In pseudocode: `hash $ prefix <> prefix' <> serialiseCIP58 n`
+
+#### Hashing negative integers
+
+The same algorithm as above is used, but the number hashed is `1 - n`.
+
+In pseudocode: `hash $ prefix <> prefix' <> serialiseCIP58 (1 - n)`
+
+#### Hashing bytestrings
+
+The bytestring is hashed as-is.
+We use the blake2b-256 hash here, such that we can usefully check that
+the script refers to a bytestring that we know only the hash of.
+
+In pseudocode: `hash $ prefix <> blake2b_256 bs`
+
+#### Hashing strings
+
+The flat-encoding is hashed.
+
+In pseudocode: `hash $ prefix <> flat x`
+
+#### Hashing lists, pairs
+
+Lists and pairs are hashed like a Merkle tree,
+much the same way that terms are.
+The children have a known type, and are hashed according to how that
+type should be hashed, i.e. with the correct algorithm and prefix.
+
+In pseudocode: `hash $ foldl' (<>) prefix (hash <$> children)`
+
+#### Hashing `()`, `False`, `True`
+
+Each has its own separate prefix, like `Error`, hence:
+
+In pseudocode: `hash prefix`
+
+#### Hashing `Data`
+
+The `CBOR` encoding is used, notably, it must be compatible with the `serialiseData`
+built-in to be useful on-chain.
+We use the blake2b-256 hash here, such that we can usefully check that
+the script refers to a datum that we know only the hash of.
+If the hashing algorithm for data changes, we must also change it here.
+
+In pseudocode: `hash $ prefix <> blake2b_256 (serialiseData d)`
+
+## Rationale
```

Comment:
We might need some discussion of the cost of this kind of hashing. Our experiments suggested it was ~10x more expensive (https://github.com/input-output-hk/plutus/blob/master/doc/notes/plutus-core/Merklisation/Merklisation-notes.md#the-cost-of-calculating-merkle-hashes), unclear if this will have a meaningful impact but it might.